### DISCUSSION TITLE
Allow custom filters on module navigation

### DISCUSSION THREAD

**Discussion by Attacler** [directus/directus#10421] [Feature Requests]
So basicly in the last 2 months+-, the module navigation per role has been removed and replaced by a "universal"/project navigation.
I would like to give a suggestion about this:
![image](https://user-images.githubusercontent.com/42644889/145442432-eb3a2b42-0a9f-4b7d-8825-bb268fecc0ea.png)
The module bar options should be clickable.
From there on you should be able to select filters for it.
For example:
Role ID equals 'abc'
or
User ID equals 'abc'

For this the normal filter rules interface should be used:
![image](https://user-images.githubusercontent.com/42644889/145442647-918d9b94-261e-44fd-85f3-247a3b7939ca.png)

I`ll throw in 30 euro if someone wants to take this on.

**Comment by benhaynes** [comment]
I wonder if it would be easier to simply duplicate the global Module setup on the role as an override. So you have the global config, and the same on the role to override... perhaps with an option to copy the default values from global to the role (as a starting point). 🤔

**Reply by Attacler** [reply]
Well yes and no.
If we move it like that it will work.
But using the filter options will basicly allow people to make it more advanced.
For example:
if user has the boolean "Subscription active" to true, they can show them differed modules than a user who has it to false.

(I know its only a frontend part and does not affect the page, you can always open the module if you know the right command)

**Reply by benhaynes** [reply]
That is true! I guess with this direction we would just need to add the conditional filters to the Module modal, right? An interesting approach. @rijkvanzanten, thoughts?

**Reply by rijkvanzanten** [reply]
Could work! It'll be the permissions based on the fields that exist in the directus_users table, as that relationally also gives you access to the role, so you could do something like

```json
{
  "role": {
    "name": {
      "_contains": "Billing"
    }
  }
}
```

**Reply by Attacler** [reply]
> 

exactly that!

**Reply by sinmsinm** [reply]
I'm very interested in this feature. We had a custom project using directus and we use administration interface to let people interact with application depending on the role.

We built a module  extension that should be only available for some roles. We are waiting for any solution like this one to replace what we had in rc90.

Congrats for the platform.

**Reply by benhaynes** [reply]
If you're interested in [sponsoring the feature](https://docs.directus.io/getting-started/support/#commissioned-features), you can contact our sales team via our website's Live Chat. ❤️

**Reply by sinmsinm** [reply]
Sure, I'll contact you. Thanks.

**Reply by Attacler** [reply]
> Sure, I'll contact you. Thanks.

Did anything happen with this feature yet?

**Comment by Pachat** [comment]
Why not just place the modules in a child collection related to `directus_settings` _instead of having them in a unique row_ ?

That collection named for instance `directus_modules` would have an M2O relationship with `directus_settings`.
That way, we could benefit of all allow accesses and filters permissions on `directus_modules` that we already have per role on collections.

**Reply by danieljwestman** [reply]
I'm upvoting this ⬆

It's been a while without the ability to customize the module bar for different users/roles and there are a lot of projects needing it.

**Comment by christianrr** [comment]
In the meantime, we were able to bypass this missing feature using a custom middleware extension:

```js
module.exports = function registerHook({ init }, { env }) {

  init('middlewares.after', async ({ app }) => {
    app.use(function (req, res, next) {
      if(req.originalUrl === '/settings') {
        const send = res.send;
        res.send = function (body) {
          let parsedBody = JSON.parse(body);
          if (parsedBody.data.module_bar) {
            if (req.accountability.role === '3f99daed-56f2-db4d-8330-sff788c5b062') {
              parsedBody.data.module_bar.find(m => m.id === 'users').enabled = false;
              parsedBody.data.module_bar.find(m => m.id === 'files').enabled = false;
            }
            else if (req.accountability.role === 'i19df5ae-4fd3-45ee-56e8-a2c932f7f8fd') {
              parsedBody.data.module_bar.find(m => m.id === 'users').enabled = false;
              parsedBody.data.module_bar.find(m => m.id === 'files').enabled = false;
            }
          }
          send.apply(this, [JSON.stringify(parsedBody)]);
        };
      }
      next();
    });
  });
};
```

No graphical editing though, but at least it allows for role-based modules in the Admin app without changing the current infrastructure.

**Comment by u12206050** [comment]
I have been trying to implement a different approach, but adding a role to the directus_settings table, and then being able to add unique project settings per role, including changing the css, available modules, project name etc. 

Upon logging in, I load the settings for the current role if there exists one else the default settings apply.

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
